packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -1,154 +1,503 @@
+  embedBatch(chunks: string[]): Promise<number[][]>;
+}
+
+export class LangChainEmbeddingWrapper implements Embedding {
+  constructor(readonly lcEmbedding: Embeddings) {}
+
+  embed(text: string): Promise<number[]> {
+    return this.lcEmbedding.embedQuery(text);
+  }
+
+  embedBatch(chunks: string[]): Promise<number[][]> {
+    return this.lcEmbedding.embedDocuments(chunks);
+  }
+}
+
+export const defaultEmbedding = new LangChainEmbeddingWrapper(new OpenAIEmbeddings());
 
 /**
- * A cast wrapper to convert an AI.JSX document to a LangChain document.
+ * An embedded chunk is a piece of a document that is ready to be added into a vector space.
  */
-export function toLangChainDoc(doc: Document): LangChainDocument {
-  return doc as LangChainDocument;
+export interface EmbeddedChunk<ChunkMetadata extends Jsonifiable = Jsonifiable> {
+  /** The identifier for this chunk. For insertions, this will be generated if not populated.*/
+  id?: string;
+
+  /** The vector representing this chunk for semantic nearest neighbors. */
+  vector: number[];
+
+  /** The content of this chunk. This is what is provided as context to an LLM when the chunk is selected. */
+  content: string;
+
+  /** The name of the document from which this chunk was extracted. */
+  documentName?: string;
+
+  /** Optional additional metadata associated with this chunk. */
+  metadata?: ChunkMetadata;
+}
+
+/** A chunk along with its score for a particular query. */
+export interface ScoredChunk<ChunkMetadata extends Jsonifiable = Jsonifiable> {
+  chunk: EmbeddedChunk<ChunkMetadata>;
+  score: number;
+}
+
+export interface Corpus<ChunkMetadata extends Jsonifiable = Jsonifiable> {
+  startCrawl: () => Promise<Corpus.Stats>;
+
+  getStats: () => Promise<Corpus.Stats>;
+
+  search: (
+    query: string,
+    params?: { limit?: number; score_threshold?: number }
+  ) => Promise<ScoredChunk<ChunkMetadata>[]>;
+}
+
+export namespace Corpus {
+  export enum LoadingState {
+    NOT_STARTED = 'NOT_STARTED',
+    IN_PROGRESS = 'IN_PROGRESS',
+    COMPLETED = 'COMPLETED',
+    FAILED = 'FAILED',
+  }
+
+  export interface Stats {
+    loadingState: LoadingState;
+    completedPartitions: number;
+    activePartitions: number;
+    numDocuments: number;
+    numChunks: number;
+  }
+}
+
+/* Resolving disagreement between formatter and linter. */
+/* eslint-disable brace-style */
+export class LocalCorpus<
+  DocumentMetadata extends Jsonifiable = Jsonifiable,
+  ChunkMetadata extends Jsonifiable = Jsonifiable
+> implements Corpus<ChunkMetadata>
+{
+  /* eslint-enable */
+  private readonly vectors: EmbeddedChunk<ChunkMetadata>[] = [];
+  private readonly documents: Document<DocumentMetadata>[] = [];
+  private readonly completedPartitions = new Set<string>();
+  private readonly activePartitionsToToken = new Map<string | null, string | null>();
+  private loadingState = Corpus.LoadingState.NOT_STARTED;
+  constructor(
+    readonly loader: CorpusLoading.Loader<DocumentMetadata>,
```

Comment:
There's syntax sugar you can use when the `constructor` is just setting instance vars:

https://www.typescriptlang.org/play?#code/MYGwhgzhAEBi0G8BQ1rAPYDsIBcBOArsDungBQqrQAOeAlgG5g4Cm0eLYAJliAJ7QARmDwAuIenQhOmSgEpEAXyTKkGbDmgAzaAF5omFgHc4ZfARZyA3EA